lib/std/terminfo.zig

```diff
@@ -0,0 +1,1122 @@
+        const FilterCallback = fn([]u8)bool;
+        const MapCallback = fn([]u8)[]const u8;
+
+        fifo: FifoType,
+        file: std.fs.File,
+        stream: Stream,
+        filter_fn: FilterCallback,
+        map_fn: MapCallback,
+
+        pub fn init(file_path: []const u8) !Self {
+            const file = try std.fs.cwd().openFile(file_path, .{});
+            var in_stream = file.inStream();
+
+            return Self {
+                .fifo = FifoType.init(),
+                .file = file,
+                .stream = in_stream.stream,
+                .filter_fn = struct { fn filter(line: []u8) bool { return true; } }.filter,
+                .map_fn = struct { fn map(line: []u8) []u8 { return line; } }.map,
+            };
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.file.close();
+        }
+
+        pub fn setFilter(self: *Self, filter_fn: FilterCallback) void {
+            self.filter_fn = filter_fn;
+        }
+
+        pub fn setMap(self: *Self, map_fn: MapCallback) void {
+            self.map_fn = map_fn;
+        }
+
+        pub fn next(self: *Self) !?[]const u8 {
+            while (try (&self.stream).readUntilDelimiterOrEof(self.fifo.buf[0..], '\n')) |line| {
+                if (self.filter_fn(line)) return self.map_fn(line);
+            }
+            return null;
+        }
+
+        pub fn collect(self: *Self) !std.Buffer {
+            var buffer = try std.Buffer.initCapacity(std.heap.page_allocator, buffer_size);
+
+            while (try self.next()) |line| {
+                try buffer.append(line);
+            }
+
+            return buffer;
+        }
+    };
+}
+
+pub const Terminfo = struct {
+    names: std.StringHashMap(bool),
+    caps: std.StringHashMap([]const u8),
+
+    pub fn init(terminfo_path: []const u8, cap_delimiter: []const u8) !Terminfo {
+        var names = std.StringHashMap(bool).init(std.heap.page_allocator);
+        var caps = std.StringHashMap([]const u8).init(std.heap.page_allocator);
+        _ = try caps.put("_name", "tui.terminfo capabilities");
+
+        var read_line = try ReadLineIterator(4096).init(terminfo_path);
+        defer read_line.deinit();
+        read_line.setFilter(struct {
+            fn filter(line: []u8) bool {
+                return !(std.mem.startsWith(u8, line, "#") or line.len == 0);
+            }
+        }.filter);
+        read_line.setMap(struct {
+            fn map(line: []u8) []const u8 {
+                return std.mem.trim(u8, line[0..], " \r\t");
+            }
+        }.map);
+
+        var buffer = try read_line.collect();
+        defer buffer.deinit();
+        var tokens = std.mem.tokenize(buffer.toSlice(), cap_delimiter);
+
+        while (tokens.next()) |token| {
+            const trimmed = std.mem.trimLeft(u8, token, " ");
+
+            if (std.mem.indexOf(u8, trimmed, "|") != null) {
+                var name_tokens = std.mem.tokenize(trimmed, "|");
```

Comment:
You could use the result of `indexof` and just grab slices on either side.

Reply:
I love the idea, however for the name tokens in particular, I saw terminfo.src examples where there could be more than 2 names for a terminal, thus with more than one `|` delimiter in between them. So I feel for just for the terminal names, `tokenize` is perfect.